### ISSUE TITLE
Add support for looping over tokens

### ISSUE THREAD

**Issue by michael-buschbeck** [michael-buschbeck/mychs-macro-magic#9]
It'd be nice to have a looping construct that executes the same block of code for all (visible) tokens on the table.

**Comment by michael-buschbeck** [comment]
Brainstorming syntax:

| Commands | Description
| -
| _!mmm_ **foreach** [**controllable**] *identifier* [**where** *expression*] | *loop over tokens with optional filter conditions*
| _!mmm_ **end foreach**


| Commands
| -
| _!mmm_ **foreach** [**controllable**] *identifier* [**in** *expression*] | *loop over items in custom expression*
| _!mmm_ **end foreach**

**Comment by michael-buschbeck** [comment]
Another thing one might want to loop over: all rows of a given character sheet table – `findattr(sender, table, col)`, which currently doesn't return anything, could return a list of all `col` values in `table`.

**Comment by phylll** [comment]
Another thing I would want to loop over are simple custom lists. I've started writing a versatile ammunition recollection script, and I'd love to loop over an initially defined (or customized) list of variable length that tells me what to look for (arrows, bolts, throwing knives, matchsticks) without the script user having to delete some of these blocks every time they install the script:

**set** ammoTypes = "arrows", "bolts"
**foreach** ammoType **in** ammoTypes [**where** *condition*]
   ... (go forth and collect)
**end foreach**

Couldn't this be generalized if you provided functions or special variables (instead of my self-defined list `ammoTypes`) to allow for dynamic access to visible tokens or return lists of table data?

**Comment by michael-buschbeck** [comment]
It's starting to look more and more to me that we don't just want arrays – what we want is *tables* with one or several columns. (Think "array of structs".)

Tables seem to be everywhere:
- Turn tracker: `token`, `label`, `value`
- Default roll template table: `label`, `value`
- _(custom)_ Gathering ammo: `ammoType`, `ammoName`, `weaponName`

Tables could also meaningfully double as key-value maps, only that you can potentially look up any table row by any column value or combination thereof (that's where the **where** clause above would become more like SQL's **where**).

**Comment by phylll** [comment]
If feasible, yes - would love to have full access to the character sheet's tables :). And overall, certainly better to have the more powerful feature than what might be spending similar amounts of your time to implement several complicated, less powerful alternatives... I guess it would be important, however, to keep a flexible syntax that allows for simplification where all the power of tables is not required, i.e. single-column lists.

**Comment by michael-buschbeck** [comment]
Alright, so, I think that we could start with something simple like this:

| Line | Command
|-|-
| 1 | _!mmm_ **foreach** variable **in** expression
| 2 | …
| 3 | _!mmm_ **end foreach**

That solves "iterating over all selected tokens" via the new implicit `selected` variable (see #28).

Add a `range(from, to)` function – or a `from ... to` operator if I feel fancy at the time (which doesn't require the expression engine require first – simple operators are as pluggable as functions) –, and we have a viable looping construct that isn't blocked by the expression engine rewrite.

(…though still no array subscript support. Introducing a new set of brackets with special functionality is a pain in the current expression engine.)

**Comment by michael-buschbeck** [comment]
Alright, `for` is in (as well as `findattr()` returning a list of attribute names in `for` context) – how significant is `range(start,end)` or `range(start,end,step)` or similar?
